src/formula/callable_objects.cpp

```diff
@@ -849,4 +860,87 @@ void safe_call_result::get_inputs(formula_input_vector& inputs) const
 	}
 }
 
+void gamestate_callable::get_inputs(formula_input_vector &inputs) const
+{
+	add_input(inputs, "turn");
+	add_input(inputs, "time_of_day");
+	add_input(inputs, "current_side");
+	add_input(inputs, "sides");
+	add_input(inputs, "units");
+	add_input(inputs, "map");
+}
+
+variant gamestate_callable::get_value(const std::string &key) const
+{
+	if(key == "turn") {
+		return variant(resources::tod_manager->turn());
+	} else if(key == "time_of_day") {
+		return variant(resources::tod_manager->get_time_of_day().id);
+	} else if(key == "current_side") {
+		return variant(resources::controller->current_side());
+	} else if(key == "sides") {
+		std::vector<variant> vars;
+		for(const auto& team : resources::gameboard->teams()) {
+			vars.emplace_back(std::make_shared<team_callable>(team));
+		}
+		return variant(vars);
+	} else if(key == "units") {
+		std::vector<variant> vars;
+		for(const auto& unit : resources::gameboard->units()) {
+			vars.emplace_back(std::make_shared<unit_callable>(unit));
+		}
+		return variant(vars);
+	} else if(key == "map") {
+		return variant(std::make_shared<gamemap_callable>(*resources::gameboard));
+	}
+
+	return variant();
+}
+
+void event_callable::get_inputs(formula_input_vector &inputs) const
+{
+	add_input(inputs, "event");
+	add_input(inputs, "event_id");
+	add_input(inputs, "event_data");
+	add_input(inputs, "loc1");
+	add_input(inputs, "unit1");
+	add_input(inputs, "attack1");
+	add_input(inputs, "loc2");
+	add_input(inputs, "unit2");
+	add_input(inputs, "attack2");
+}
+
+variant event_callable::get_value(const std::string &key) const
+{
+	if(key == "event") {
+		return variant(event_info.name);
+	} else if(key == "event_id") {
+		return variant(event_info.id);
+	} else if(key == "loc1") {
+		return variant(std::make_shared<location_callable>(event_info.loc1));
+	} else if(key == "loc2") {
+		return variant(std::make_shared<location_callable>(event_info.loc2));
+	} else if(key == "event_data") {
+		return variant(std::make_shared<config_callable>(event_info.data));
+	} else if(key == "unit1") {
+		if(auto u1 = event_info.loc1.get_unit()) {
+			return variant(std::make_shared<unit_callable>(*u1));
+		}
+	} else if(key == "unit2") {
+		if(auto u2 = event_info.loc2.get_unit()) {
+			return variant(std::make_shared<unit_callable>(*u2));
+		}
+	} else if(key == "attack1") {
+		if(event_info.data.has_child("first")) {
+			return variant(std::make_shared<attack_type_callable>(attack_type(event_info.data.child("first"))));
+		}
+	} else if(key == "attack2") {
```

Comment:
Is this creating a set of new names for existing concepts, similar to "repeating" vs "first_time_only"?

Reply:
Maybe? I guess in wesnoth.current they're called `weapon` and `second_weapon`…

Reply:
I think the API should be consistent, as discussed in https://www.wesnoth.org/irclogs/2021/04/%23wesnoth-dev.2021-04-11.log between 12:00 and 19:45.

While searching for an example I noticed there was no unit test for formulas using these values.

Reply:
Please don't use different names for existing functionality and concepts - even though this particular functionality is new, it doesn't exist in a vacuum, and should follow existing names even if you don't think they're completely ideal.

Otherwise in a few years someone else is probably going to try and propose changing these new names to be consistent with the existing names anyway, and I don't think any UMC authors are going to be happy to deal with the names starting off different between the APIs and then anyone who uses to new API having to do a bunch of boilerplate work to change the names later.

Reply:
> While searching for an example I noticed there was no unit test for formulas using these values.

Okay, added. It's a good thing you mentioned this, because those values actually didn't work due to them not being valid WFL identifiers.